Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,314 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP1
+import NIOSSL
+import NIOTLS
+#if canImport(Network)
+    import NIOTransportServices
+#endif
+
+extension HTTPConnectionPool {
+    enum NegotiatedProtocol {
+        case http1_1(Channel)
+        case http2_0(Channel)
+    }
+
+    final class ConnectionFactory {
+        let key: ConnectionPool.Key
+        let clientConfiguration: HTTPClient.Configuration
+        let tlsConfiguration: TLSConfiguration
+        let sslContextCache: SSLContextCache
+
+        init(key: ConnectionPool.Key,
+             tlsConfiguration: TLSConfiguration?,
+             clientConfiguration: HTTPClient.Configuration,
+             sslContextCache: SSLContextCache) {
+            self.key = key
+            self.clientConfiguration = clientConfiguration
+            self.sslContextCache = sslContextCache
+            self.tlsConfiguration = tlsConfiguration ?? clientConfiguration.tlsConfiguration ?? .forClient()
+        }
+    }
+}
+
+extension HTTPConnectionPool.ConnectionFactory {
+    func makeBestChannel(connectionID: HTTPConnectionPool.Connection.ID, eventLoop: EventLoop, logger: Logger) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        if self.key.scheme.isProxyable, let proxy = self.clientConfiguration.proxy {
+            return self.makeHTTPProxyChannel(proxy, connectionID: connectionID, eventLoop: eventLoop, logger: logger)
+        } else {
+            return self.makeChannel(eventLoop: eventLoop, logger: logger)
+        }
+    }
+
+    private func makeChannel(eventLoop: EventLoop, logger: Logger) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        switch self.key.scheme {
+        case .http, .http_unix, .unix:
+            return self.makePlainChannel(eventLoop: eventLoop).map { ($0, .http1_1) }
+        case .https, .https_unix:
+            return self.makeTLSChannel(eventLoop: eventLoop, logger: logger).map {
+                (channel, negotiated) -> (Channel, HTTPVersion) in
+                let version = negotiated == "h2" ? HTTPVersion.http2 : HTTPVersion.http1_1
```

Comment:
Something feels off about using a string `h2` to identify the channel type here. would an enum be better? Or maybe we could switch `negotiated` to account for `H2` or something?

Reply:
Yep this still applies @fabianfett